src/test/java/org/prlprg/rds/RDSWriterTest.java

```diff
@@ -0,0 +1,428 @@
+      }
+    } else {
+      fail("Expected VecSXP");
+    }
+  }
+
+  @Test
+  public void testList() throws Exception {
+    var elems =
+        new TaggedElem[] {
+          new TaggedElem("a", SEXPs.integer(1)),
+          new TaggedElem("b", SEXPs.logical(Logical.TRUE)),
+          new TaggedElem("c", SEXPs.real(3.14, 2.71))
+        };
+    var list = SEXPs.list(elems, Attributes.NONE);
+    var output = new ByteArrayOutputStream();
+
+    RDSWriter.writeStream(output, list);
+
+    var input = new ByteArrayInputStream(output.toByteArray());
+    var sexp = RDSReader.readStream(rsession, input);
+
+    if (sexp instanceof ListSXP l) {
+      assertEquals(3, l.size());
+      assertEquals("a", l.get(0).tag());
+      if (l.get(0).value() instanceof IntSXP i) {
+        assertEquals(1, i.get(0));
+      } else {
+        fail("Expected IntSXP for the 1st element of the ListSXP");
+      }
+      assertEquals("b", l.get(1).tag());
+      if (l.get(1).value() instanceof LglSXP lgl) {
+        assertEquals(Logical.TRUE, lgl.get(0));
+      } else {
+        fail("Expected LglSXP for the 2nd element of the ListSXP");
+      }
+      assertEquals("c", l.get(2).tag());
+      if (l.get(2).value() instanceof RealSXP r) {
+        assertEquals(3.14, r.get(0));
+        assertEquals(2.71, r.get(1));
+      } else {
+        fail("Expected RealSXP for the 3rd element of the ListSXP");
+      }
+    } else {
+      fail("Expected ListSXP");
+    }
+  }
+
+  @Test
+  public void testEnv() throws Exception {
+    var env = new UserEnvSXP();
+    env.set("a", SEXPs.integer(1));
+    env.set("b", SEXPs.logical(Logical.TRUE));
+    env.set("c", SEXPs.real(3.14, 2.71));
+    env.set("d", SEXPs.string("foo", "bar"));
+    env.setAttributes(new Attributes.Builder().put("test", SEXPs.logical(Logical.TRUE)).build());
+
+    var output = new ByteArrayOutputStream();
+
+    RDSWriter.writeStream(output, env);
+
+    var input = new ByteArrayInputStream(output.toByteArray());
+    var sexp = RDSReader.readStream(rsession, input);
+
+    if (sexp instanceof UserEnvSXP read_env) {
+      assertEquals(4, read_env.size());
+      assertEquals(SEXPs.integer(1), read_env.get("a").orElseThrow());
+      assertEquals(SEXPs.logical(Logical.TRUE), read_env.get("b").orElseThrow());
+      assertEquals(SEXPs.real(3.14, 2.71), read_env.get("c").orElseThrow());
+      assertEquals(SEXPs.string("foo", "bar"), read_env.get("d").orElseThrow());
+      assertEquals(
+          new Attributes.Builder().put("test", SEXPs.logical(Logical.TRUE)).build(),
+          read_env.attributes());
+    } else {
+      fail("Expected UserEnvSXP");
+    }
+  }
+
+  @Test
+  public void testEnv_withR() throws Exception {
+    var env = new UserEnvSXP();
+    env.set("a", SEXPs.integer(1));
+    env.set("b", SEXPs.logical(Logical.TRUE));
+    env.set("c", SEXPs.real(3.14, 2.71));
+    env.set("d", SEXPs.string("foo", "bar"));
+
+    var output =
+        R.eval(
```

Comment:
I don't know how to format this well, but we should have all the multiline strings formatted the same way.